core/modules/user/user.module

```diff
@@ -1131,7 +1129,15 @@ function user_edit_access($account) {
  * users, and prevent the anonymous user from cancelling the account.
  */
 function user_cancel_access($account) {
-  return ((($GLOBALS['user']->uid == $account->uid) && user_access('cancel account')) || user_access('administer users')) && $account->uid > 0;
+  if (!($account instanceof User) && is_int($account)) {
+    $account = user_load($account);
+  }
```

Comment:
Same here, no need to do the integer check here as the old function did not allow cancel checking access on integers.